### PR TITLE
conversion v1 to v3 node

### PR DESCRIPTION
**Pull request by tmjd**
```release-note
None required
```

### CODE REVIEW

File: lib/upgrade/etcd/conversionv1v3/node.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+		}
+
+		// If the backend has an IPv4 address then fill in the IPv4Address
+		// field.  If the IP network does not exist assume a full mask.
+		if bv.BGPIPv4Addr != nil {
+			if bv.BGPIPv4Net != nil {
+				// Stored network is normalised, so copy across the
+				// IP separately.
+				apiNode.Spec.BGP.IPv4Address = bv.BGPIPv4Net.String()
+				//apiNode.Spec.BGP.IPv4Address.IP = bv.BGPIPv4Addr.IP
```
Comment (caseydavenport) [inline lib/upgrade/etcd/conversionv1v3/node.go]: Commented out code still here.

File: lib/upgrade/etcd/conversionv1v3/node.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	bk, ok := d.Key.(model.NodeKey)
+	if !ok {
+		return nil, fmt.Errorf("Key is not a valid NodeKey resource: %v", d.Value)
+	}
+
+	apiNode := apiv3.NewNode()
+
+	apiNode.ObjectMeta.Name = convertName(bk.Hostname)
```
Comment (caseydavenport) [inline lib/upgrade/etcd/conversionv1v3/node.go]: We'll need to use the `convertNodeName` function when it exists.

File: lib/upgrade/etcd/conversionv1v3/node.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+)
+
+type Node struct{}
+
+// convertKVPairToAPI converts a KVPair containing a backend Node and NodeKey
+// to an API Node structure.
+// This is part of the conversionHelper interface.
+func (_ Node) BackendV1ToAPIV3(d *model.KVPair) (Resource, error) {
```
Comment (caseydavenport) [inline lib/upgrade/etcd/conversionv1v3/node.go]: What's the plan for populating `Node.Spec.NodeBGPSpec.IPv4IPIPTunnelAddr`?

@tmjd @robbrockbank?

Comment (robbrockbank) [reply lib/upgrade/etcd/conversionv1v3/node.go]: That'll be retro-fitted by the upgrade script, but for simple YAML conversion it won't be possible to include that field.

Comment (tmjd) [reply lib/upgrade/etcd/conversionv1v3/node.go]: I added a comment saying that it needs to be populated external to the function.

File: lib/upgrade/etcd/conversionv1v3/node.go (hunk 2 above)
Comment (tmjd) [reply lib/upgrade/etcd/conversionv1v3/node.go]: See if you like what I did with the alias function.

File: lib/upgrade/etcd/conversionv1v3/names.go [hunk 4]
```diff
@@ -66,6 +66,11 @@ func convertNameNoDots(v1Name string) string {
 	return name
 }
 
+// Currently just an alias for convertName but it is expected to be different
```
Comment (caseydavenport) [inline lib/upgrade/etcd/conversionv1v3/names.go]: I think we probably don't want to do this aliasing - once we have a function we'll go around and find all the places we need to update.

For now, let's just use `convertName` in line.

**Comment by tmjd** [comment]
I think this is ready to go now.

**Pull request merged**